### ISSUE TITLE
False positive on torch.no_grad decorator

### ISSUE THREAD

**Issue by ltskv** [astral-sh/ty#2594]
### Summary

Hi! Thanks for your amazing tooling! Encountered this false positive when trying to apply `ty` to our codebase.

Minimal example:

```python
import torch


@torch.no_grad
def stuff(x: torch.Tensor) -> torch.Tensor:
    return x + 1
```

gives

```
error[too-many-positional-arguments]: Too many positional arguments to bound method `__init__`: expected 1, got 2
 --> stuff.py:4:1
  |
4 | @torch.no_grad
  | ^^^^^^^^^^^^^^
5 | def stuff(x: torch.Tensor) -> torch.Tensor:
6 |     return x + 1
  |
info: Method signature here
  --> /home/pavel/.cache/uv/archive-v0/AQ5UTMgr1gW-KMsLkcacH/lib/python3.12/site-packages/torch/autograd/grad_mode.py:75:9
   |
73 |     """
74 |
75 |     def __init__(self) -> None:
   |         ^^^^^^^^^^^^^^^^^^^^^^
76 |         if not torch._jit_internal.is_scripting():
77 |             super().__init__()
   |
info: rule `too-many-positional-arguments` is enabled by default
```

This diagnostic can be a bit cryptic for an unprepared developer :D

The parent class of `no_grad` has this implementaiton for `__new__`, so during runtime the `__init__` in question is never called with arguments.

```python
class no_grad(_NoParamDecoratorContextManager):
    ...

class _NoParamDecoratorContextManager(_DecoratorContextManager):
    """Allow a context manager to be used as a decorator without parentheses."""

    def __new__(cls, orig_func=None):
        if orig_func is None:
            return super().__new__(cls)
        return cls()(orig_func)
```

### Version

ty 0.0.13

**Comment by carljm** [comment]
Thanks for reporting this!

I think this is a duplicate of #281 -- if we recognize that `__new__` may return some other type, we should not check the call against `__init__` signature.

**Comment by carljm** [comment]
Confirmed fixed on main (not yet released).
